File: Problem3.java
import java.util.Arrays;

class Solution {
    public String[] reorderLogFiles(String[] logs) {
        Arrays.sort(logs, (a, b) -> {
            String[] splitA = a.split(" ", 2);
            String[] splitB = b.split(" ", 2);
            
            boolean isDigit1 = Character.isDigit(splitA[1].charAt(0));
            boolean isDigit2 = Character.isDigit(splitB[1].charAt(0));
            
            if (!isDigit1 && !isDigit2) {
                int cmp = splitA[1].compareTo(splitB[1]);
                if (cmp != 0) {
                    return cmp;
                }
                return splitA[0].compareTo(splitB[0]);
            } else if (isDigit1 && !isDigit2) {
                return 1; // 'a' is a digit log, 'b' is a letter log
            } else if (!isDigit1 && isDigit2) {
                return -1; // 'a' is a letter log, 'b' is a digit log
            } else {
                return 0; // both logs are digit logs
            }
        });
        return logs;
    }
}

File: Problem3.py
class Solution:
    def reorderLogFiles(self, logs: List[str]) -> List[str]:
        '''
        Time Complexity: O(N * M * log N), custom sorting function -->O(M) time to split the log string, where M is the average length of the logs. Sorting N logs using this function takes O(N * log N) time. 
        Space Complexity:  O(N) storage of the reordered logs.
        '''
        # Custom sorting function to determine the order of logs
        def custom_sort(log):
            identifier, content = log.split(" ", 1)
            # If the content starts with a digit, it's a digit-log
            if content[0].isdigit():
                return (1, )  # Append 1 to push digit-logs to the end
            else:
                # Sort letter-logs lexicographically
                return (0, content, identifier)

        # Sort the logs using the custom sorting function
        logs.sort(key=custom_sort)
        return logs
